File: ProgressBar.py
import sys
import math
import os
import unicodedata
import colorama
from colorama import Fore,Back,Style
colorama.init()

#モノクロ版プログレスバー
class Progress:
	def __init__(self,width,parMax):
		#最大の行数
		self.columns=os.get_terminal_size().columns
		#プログレスバーの長さ
		self.width=width
		#進捗度
		self.par=0
		#目標進捗度
		self.parMax=parMax

	#プログレスバーの更新
	def update(self,message):
		parcent=self.par/self.parMax
		widthNow=math.floor(self.width*parcent)
		rowCnt=math.floor(self.__lenB(message)/self.columns)+2

		gauge=">"*widthNow+" "*(self.width-widthNow)
		status=f"({round(parcent*100,1)}%<-{self.par}/{self.parMax})"

		sys.stderr.write(
			f"#[{gauge}]#{status}\n"+
			"\u001b[0J"+			#clearScreenDown
			message+
			f"\u001b[{rowCnt}A\r\n" #moveCursorUp
		)
		self.par=self.par+1

	#プログレスバーの完了
	def done(self,doneAlert):
		sideLen=math.floor((self.width-len(doneAlert))/2)
		gauge="="*sideLen+doneAlert
		gauge+="="*(self.width-len(gauge))

		status=f"(100%<-{self.parMax}/{self.parMax})"

		sys.stderr.write(
			"\u001b[0J"+	#clearScreenDown
			f"#[{gauge}]#{status}\n"
		)

	#バイト(文字幅)数計算
	def __lenB(self,str):
		len=0
		for val in str:
			cw=unicodedata.east_asian_width(val)
			if cw in u"WFA":
				len=len+2
			else:
				len=len+1
		return len

#カラー版プログレスバー
class ProgressColor(Progress):
	#プログレスバーの更新
	def update(self,message):
		parcent=self.par/self.parMax
		widthNow=math.floor(self.width*parcent)
		rowCnt=math.floor(self._Progress__lenB(message)/self.columns)+2

		status=f"({round(parcent*100,1)}%<-{self.par}/{self.parMax})"

		sys.stderr.write(
			Style.BRIGHT+
			Back.YELLOW+
			Fore.YELLOW+
			"{"+
			Back.CYAN+
			Fore.CYAN+
			"▤"*widthNow+
			" "*(self.width-widthNow)+
			Back.YELLOW+
			Fore.YELLOW+
			"}"+
			Style.RESET_ALL+
			status+
			"\n"+
			"\u001b[0J"+			#clearScreenDown
			message+
			f"\u001b[{rowCnt}A\r\n" #moveCursorUp
		)
		self.par=self.par+1

	#プログレスバーの完了
	def done(self,doneAlert):
		sideLen=math.floor((self.width-len(doneAlert))/2)
		doneAlert=" "+doneAlert+" "
		gauge=(
			Fore.GREEN+
			"▤"*sideLen+
			Fore.RED+
			doneAlert+
			Fore.GREEN+
			"▤"*(self.width-len("#"*sideLen+doneAlert))
		)
		status=f"(100%<-{self.parMax}/{self.parMax})"

		sys.stderr.write(
			"\u001b[0J"+	#clearScreenDown
			Style.BRIGHT+
			Back.YELLOW+
			Fore.YELLOW+
			"{"+
			Back.GREEN+
			gauge+
			Back.YELLOW+
			Fore.YELLOW+
			"}"+
			Style.RESET_ALL+
			status+
			"\n"
		)

File: ProgressBar.js
"use strict";
const readline=require("readline");

const ProgressBar=(()=>{
	const PB={};
	const lenB=Symbol();

	//モノクロ版プログレスバー
	PB.Progress=class{
		constructor(width,parMax){
			//最大の桁数
			this.columns=process.stdout.columns;
			//プログレスバーの長さ
			this.width=width;
			//進捗度
			this.par=0;
			//目標進捗度
			this.parMax=parMax;
		}

		//プログレスバーの更新
		update(message){
			const parcent=this.par/this.parMax;
			const widthNow=Math.floor(this.width*parcent);
			const rowCnt=Math.floor(this[lenB](message)/this.columns)+2;

			const gauge=">".repeat(widthNow)+" ".repeat(this.width-widthNow);
			const status=`(${(parcent*100).toFixed(1)}%<-${this.par}/${this.parMax})`;

			console.error(`#[${gauge}]#${status}`);
			readline.clearScreenDown(process.stdout);

			console.error(message);
			readline.moveCursor(process.stdout,0,-rowCnt);
			this.par++;
		}

		//プログレスバーの完了
		done(doneAlert){
			const sideLen=Math.floor((this.width-doneAlert.length)/2);

			var gauge="=".repeat(sideLen)+doneAlert;
			gauge+="=".repeat(this.width-gauge.length);
			const status=`(100%<-${this.parMax}/${this.parMax})`;

			readline.clearScreenDown(process.stdout);
			console.error(`#[${gauge}]#${status}`);
		}
	};
	//バイト(文字幅)数計算
	PB.Progress.prototype[lenB]=function(message){
		var len=0;
		for(let val of message){
			let cc=val.charCodeAt(0);
			if( 0x0000<=cc && cc<=0x024F &&
				cc!=0x0085 && cc!=0x089 && cc!=0x00A7 && cc!=0x00B0 &&
				cc!=0x00B1 && cc!=0x00D7 && cc!=0x00F7 ||
				cc==0xA5 || cc==0x203E || cc==0xF8F0 ||
				0xFF61<=cc && cc<=0xFFDC ||
				0xFFE8<=cc && cc<=0xFFEE
			){
				len=0|len+1;
			}
			else{
				len=0|len+2;
			}
		}
		return len;
	};

	//カラー版プログレスバー
	PB.ProgressColor=class extends PB.Progress{
		//プログレスバーの更新
		update(message){
			const parcent=this.par/this.parMax;
			const widthNow=Math.floor(this.width*parcent);
			const rowCnt=Math.floor(this[lenB](message)/this.columns)+2;

			const status=`(${(parcent*100).toFixed(1)}%<-${this.par}/${this.parMax})`;

			readline.clearScreenDown(process.stderr);
			console.error(
				"\u001b[43m\u001b[5m"+  /*BackLightYellow*/
				"\u001b[33m"+		   /*ForeDarkYellow*/
				"{"+
				"\u001b[46m"+		   /*BackLightCyan*/
				" ".repeat(widthNow)+
				"\u001b[25m"+		   /*BackDarkCyan*/
				" ".repeat(this.width-widthNow)+
				"\u001b[43m\u001b[5m"+  /*BackLightYellow*/
				"}"+
				"\u001b[0m"+			/*ResetColor*/
				status+
				"\n"+
				message
			);
			readline.moveCursor(process.stderr,0,-rowCnt);
			this.par=0|this.par+1
		}

		//プログレスバーの完了
		done(doneAlert){
			const sideLen=Math.floor((this.width-doneAlert.length)/2);

			var gauge=" ".repeat(sideLen)+doneAlert;
			gauge+=" ".repeat(this.width-gauge.length);
			const status=`(100%<-${this.parMax}/${this.parMax})`;

			readline.clearScreenDown(process.stderr);

			console.error(
				"\u001b[43m\u001b[5m"+  /*BackLightYellow*/
				"\u001b[33m"+		   /*ForeDarkYellow*/
				"{"+
				"\u001b[42m"+		   /*BackLightGreen*/
				"\u001b[31m\u001b[1m"+  /*ForeLightRed*/
				gauge+
				"\u001b[43m"+		   /*BackLightYellow*/
				"\u001b[33m\u001b[22m"+ /*ForeDarkYellow*/
				"}"+
				"\u001b[0m"+			/*ColorReset*/
				status
			);
		}
	};

	return PB;
})();
Object.freeze(ProgressBar);
module.exports=ProgressBar;